rustc: keep upvars tupled in {Closure,Generator}Substs.

eddyb opened the pull request:
Previously, each closure/generator capture's (aka "upvar") type was tracked as one "synthetic" type parameter in the closure/generator substs, and figuring out where the parent `fn`'s generics end and the synthetics start involved slicing at `tcx.generics_of(def_id).parent_count`.

Needing to query `generics_of` limited @davidtwco (who wants to compute some `TypeFlags` differently for parent generics vs upvars, and `TyCtxt` is not available there), which is how I got started on this, but it's also possible that the `generics_of` queries are slowing down `{Closure,Generator}Substs` methods.

To give an example, for a `foo::<T, U>::{closure#0}` with captures `x: X` and `y: Y`, substs are:
* before this PR: `[T, U, /*kind*/, /*signature*/, X, Y]`
* after this PR: `[T, U, /*kind*/, /*signature*/, (X, Y)]`

You can see that, with this PR, no matter how many captures, the last 3 entries in the substs (or 5 for a generator) are always the "synthetic" ones, with the last one being the tuple of capture types.

r? @nikomatsakis cc @Zoxc

eddyb commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit da3090785790aa3ea0aa0ba0272802bdd7d5ea30 with merge 6429062cf537bcdeb8a9d276f4136375635d6e4d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"da3090785790aa3ea0aa0ba0272802bdd7d5ea30","merge_sha":"6429062cf537bcdeb8a9d276f4136375635d6e4d"} -->

eddyb commented on src/test/ui/issues/issue-22638.stderr:
```diff
@@ -8,7 +8,7 @@ LL | |         a.matches(f)
    | |_____^
    |
-   = note: consider adding a `#![type_length_limit="26214380"]` attribute to your crate
+   = note: consider adding a `#![type_length_limit="30408681"]` attribute to your crate
```
@rust-lang/compiler This is a bit silly, can we remove `type_length_limit`?
Does it do anything for us nowadays? Other than slow down compilation.

Centril replied:
I assume by "remove" you mean "make it do nothing"?

eddyb replied:
Oh, right, it's stable. The other thing we can do is count only unique types, which should make cases like these far less drastic presumably.

nikomatsakis replied:
I'm confused -- why would it not do something for us? There are various cases where the size of types and amount of computational work can balloon without generating large stacks (also true: in the SLG solver Chalk at least, we don't even have a concept of overflow apart from type-size limits, though depending on what precise solver design we wind up with, we could keep something like overflow, but it's not a great concept because it makes results for a given query dependent on context instead of independent).

nikomatsakis replied:
That said, this is an orthogonal place where we could apply type-limits, instead of applying them at type creation time or whatever.

eddyb replied:
I think it was added originally to not generate LLVM type names that were too long or something, which is a silly problem IMO.

At the very least we should switch to a DAG approach (i.e. don't count duplicates), I assume Chalk at least is similarly proportional in unique types, not total occurrences.

nikomatsakis replied:
I'm not sure what duplicates exactly means? Oh, I guess like `Foo<(A, A)>` only counts `A` once? Chalk is currently just counting the total number of types I think. I'd have to think if the DAG version "suffices". I'm not sure, why is that much better? I'm just a bit confused about why you care about this I guess. Is it something that you've seen arise frequently?

eddyb replied:
> Is it something that you've seen arise frequently?

It's almost *guaranteed* to arise with closures in generic functions, the generic shows up once in the parent `Substs` and then in one or more capture types.

That multiplies during monomorphization, and can blow out of proportions, *despite* nothing actually operating proportionally to the multiple occurrences.

Remember #54540? :rofl:

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

eddyb commented on src/librustc_typeck/check/closure.rs:
```diff
@@ -92,31 +92,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 .into(),
             GenericParamDefKind::Const => span_bug!(expr.span, "closure has const param"),
         });
+        let tupled_upvars =
```
There's a comment a bit above, I'll have to update it to mention synthetics I guess.

Centril replied:
I feel like it would be helpful to have a comment here as well, with a more precise note about the tuple.

eddyb replied:
Right, sorry, I neant update that comment to not be confusing given a new comment here.

eddyb commented:
@bors try

bors commented:
:hourglass: Trying commit da3090785790aa3ea0aa0ba0272802bdd7d5ea30 with merge 26cab41b863fc625c0d81f771e063c685680f335...
<!-- homu: {"type":"TryBuildStarted","head_sha":"da3090785790aa3ea0aa0ba0272802bdd7d5ea30","merge_sha":"26cab41b863fc625c0d81f771e063c685680f335"} -->

oli-obk commented on src/librustc/ty/sty.rs:
```diff
@@ -475,18 +470,21 @@ impl<'tcx> GeneratorSubsts<'tcx> {
+            Tuple(upvars) => upvars,
+            _ => bug!("upvars should be tupled"),
+        };
+        upvars.iter().map(|t| {
```
```suggestion
        self.split().tupled_upvars_ty.tuple_fields().map(|t| {
```

eddyb replied:
Considered that and `expect_ty`, worried it was less informative if it failed.

oli-obk replied:
you can slap `#[track_caller]` on the two functions, that should mitigate that, right?

oli-obk commented on src/librustc/ty/sty.rs:
```diff
@@ -631,16 +629,16 @@ pub enum UpvarSubsts<'tcx> {
+            _ => bug!("upvars should be tupled"),
         };
-        upvar_kinds.iter().map(|t| {
+        upvars.iter().map(|t| {
```
Same, here, use [tuple_fields](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.tuple_fields)

oli-obk commented on src/librustc/ty/sty.rs:
```diff
@@ -631,16 +629,16 @@ pub enum UpvarSubsts<'tcx> {
         };
-        upvar_kinds.iter().map(|t| {
+        upvars.iter().map(|t| {
             if let GenericArgKind::Type(ty) = t.unpack() {
```
The entire closure body can be `t.expect_ty()`

eddyb replied:
Pre-existing, I wonder what @nikomatsakis and @Zoxc think about it.

Zoxc replied:
You can probably just call `tuple_fields` here. Not sure why we use `SubstsRef` for tuples.

varkor replied:
`SubstsRef` is used for tuples due to the refactoring in https://github.com/rust-lang/rust/issues/42340.

oli-obk commented on src/librustc/ty/sty.rs:
```diff
@@ -475,18 +470,21 @@ impl<'tcx> GeneratorSubsts<'tcx> {
+            _ => bug!("upvars should be tupled"),
+        };
+        upvars.iter().map(|t| {
             if let GenericArgKind::Type(ty) = t.unpack() {
```
The entire closure body can be `t.expect_ty()`

oli-obk commented on src/librustc/ty/sty.rs:
```diff
@@ -372,31 +366,33 @@ pub struct ClosureSubsts<'tcx> {
-        let SplitClosureSubsts { upvar_kinds, .. } = self.split(def_id, tcx);
-        upvar_kinds.iter().map(|t| {
+    pub fn upvar_tys(self) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
+        let upvars = match self.tupled_upvars_ty().kind {
```
`tuple_fields` + `expect_ty` changes apply here, too.

davidtwco commented on src/librustc/ty/sty.rs:
```diff
@@ -372,31 +366,33 @@ pub struct ClosureSubsts<'tcx> {
-    fn split(self, def_id: DefId, tcx: TyCtxt<'_>) -> SplitClosureSubsts<'tcx> {
-        let generics = tcx.generics_of(def_id);
-        let parent_len = generics.parent_count;
+    fn split(self) -> SplitClosureSubsts<'tcx> {
```
I've already mentioned this to @eddyb via PM, but it could be useful for `SplitClosureSubsts` to contain a `parent_substs: &'tcx [GenericArg<'tcx>]` field - that way any callers that need to access the specific components of the closure substs (and use `SplitClosureSubsts` to do so) can also access the parent substs without having to reproduce some of this logic.

eddyb replied:
You can add this in your PR, FWIW, but if @nikomatsakis doesn't mind I can do it in this one too.

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

eddyb commented:
@bors try (CI outage should be fixed now)

bors commented:
:hourglass: Trying commit da3090785790aa3ea0aa0ba0272802bdd7d5ea30 with merge f7b1e868f01ad49dcc05daad583793f5c413bc9a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"da3090785790aa3ea0aa0ba0272802bdd7d5ea30","merge_sha":"f7b1e868f01ad49dcc05daad583793f5c413bc9a"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=23140)
Build commit: f7b1e868f01ad49dcc05daad583793f5c413bc9a (`f7b1e868f01ad49dcc05daad583793f5c413bc9a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=23140"},"merge_sha":"f7b1e868f01ad49dcc05daad583793f5c413bc9a"} -->

rust-timer commented:
Queued f7b1e868f01ad49dcc05daad583793f5c413bc9a with parent 54b7d21f59a363e53eb1c31d76b40af2ff99321c, future [comparison URL](https://perf.rust-lang.org/compare.html?start=54b7d21f59a363e53eb1c31d76b40af2ff99321c&end=f7b1e868f01ad49dcc05daad583793f5c413bc9a).

eddyb commented:
* `inflate-check` is the winner here, ~1.2% improvement
* "incremental clean" runs are all slower, [`unused-warnings-check`'s](https://perf.rust-lang.org/detailed-query.html?commit=f7b1e868f01ad49dcc05daad583793f5c413bc9a&base_commit=54b7d21f59a363e53eb1c31d76b40af2ff99321c&benchmark=unused-warnings-check&run_name=clean%20incremental) is the worst I can find

I think the regression is the "Loading: 0.012" in that `unused-warnings-check` run.
And it's from... `predicates_of`, `generics_of` and `codegen_fn_attrs`?

That's not what I expected. I was thinking maybe there is a slowdown because of the extra tupling, but it looks like it's more from. Are there enough closures w/o upvars that this PR increases the number of synthetic generics for, from 2 to 3?

This makes me want to revise my strategy and just tuple *all* synthetics together, not just upvars.

eddyb commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit a43feadf0dd2ed614d780533f73ae355de5af4f6 with merge 5774a659136586220cc1baba127f228be95ee2b1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a43feadf0dd2ed614d780533f73ae355de5af4f6","merge_sha":"5774a659136586220cc1baba127f228be95ee2b1"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=23332)
Build commit: 5774a659136586220cc1baba127f228be95ee2b1 (`5774a659136586220cc1baba127f228be95ee2b1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=23332"},"merge_sha":"5774a659136586220cc1baba127f228be95ee2b1"} -->

rust-timer commented:
Queued 5774a659136586220cc1baba127f228be95ee2b1 with parent 5a72ecf2c5c97933cecce91cb58d104da8120624, future [comparison URL](https://perf.rust-lang.org/compare.html?start=5a72ecf2c5c97933cecce91cb58d104da8120624&end=5774a659136586220cc1baba127f228be95ee2b1).

rust-timer commented:
Finished benchmarking try commit 5774a659136586220cc1baba127f228be95ee2b1, [comparison URL](https://perf.rust-lang.org/compare.html?start=5a72ecf2c5c97933cecce91cb58d104da8120624&end=5774a659136586220cc1baba127f228be95ee2b1).

eddyb commented:
My analysis in https://github.com/rust-lang/rust/pull/69968#issuecomment-599049005 was bad, I was misreading the "Loading" part of the table.
By the looks of things, tupling all synthetics is overall slower than doing it just for upvars.

I suppose overall the profiles still don't make much sense.
Take [`unused-warning-check`'s "clean incremental"](https://perf.rust-lang.org/detailed-query.html?commit=f7b1e868f01ad49dcc05daad583793f5c413bc9a&base_commit=54b7d21f59a363e53eb1c31d76b40af2ff99321c&benchmark=unused-warnings-check&run_name=clean%20incremental) from the original perf run, one of the worst regressions. But the query view claims a delta of `-0.007`!

But if we switch to [`cpu-clock`](https://perf.rust-lang.org/compare.html?start=54b7d21f59a363e53eb1c31d76b40af2ff99321c&end=f7b1e868f01ad49dcc05daad583793f5c413bc9a&stat=cpu-clock) (again, this is the original perf run, not the second one), we can see a `-1.8%` improvement instead of a `1.4%` regression.

So the default view is `instructions:u` but the query profiling data is `cpu-clock`.
The former is less noisy AFAICT, but we don't support using instruction counts in query profiling.

cc @Mark-Simulacrum @wesleywiser Do you have any advice?

Even if I can't rely on any profiling data, I might have a few things to try wrt `{Closure,Generator}Substs` that might speed up operations on them.

eddyb commented:
Added a commit that defers `GenericArg::expect_ty` as much as possible, to avoid duplicate unused work due to how `.split()` is used on `{Closure,Generator}Substs`.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 814330f90be83c7d98eef48b7e30c486694b5c9b with merge 81fe3b5cf76ef55455b4074dc863b99904f09fb4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"814330f90be83c7d98eef48b7e30c486694b5c9b","merge_sha":"81fe3b5cf76ef55455b4074dc863b99904f09fb4"} -->

bors commented:
:sunny: Try build successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=23569)
Build commit: 81fe3b5cf76ef55455b4074dc863b99904f09fb4 (`81fe3b5cf76ef55455b4074dc863b99904f09fb4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=23569"},"merge_sha":"81fe3b5cf76ef55455b4074dc863b99904f09fb4"} -->

rust-timer commented:
Queued 81fe3b5cf76ef55455b4074dc863b99904f09fb4 with parent 7ceebd98c6a15ae30e772eebb676b63ffa1a8122, future [comparison URL](https://perf.rust-lang.org/compare.html?start=7ceebd98c6a15ae30e772eebb676b63ffa1a8122&end=81fe3b5cf76ef55455b4074dc863b99904f09fb4).

rust-timer commented:
Finished benchmarking try commit 81fe3b5cf76ef55455b4074dc863b99904f09fb4, [comparison URL](https://perf.rust-lang.org/compare.html?start=7ceebd98c6a15ae30e772eebb676b63ffa1a8122&end=81fe3b5cf76ef55455b4074dc863b99904f09fb4).

eddyb commented:
Just removed the `[HACK] rustc: tuple all closure/generator synthetics, not just upvars.` commit and its revert commit, the combined diff should remain the same.
Now that perf is more reasonable, we should be able to land this.

The optimization commit addressed @oli-obk's `expect_ty`/`tuple_fields` concerns, and I'll soon add another that should address @Centril's concern about the comment in `check::closure`.

bors commented:
:umbrella: The latest upstream changes (presumably #70118) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 7c3902193ec06b19c5eb406089890ceb30a25080 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 7c3902193ec06b19c5eb406089890ceb30a25080 -->
<!-- homu: {"type":"Approved","sha":"7c3902193ec06b19c5eb406089890ceb30a25080","approver":"nikomatsakis"} -->

eddyb commented on src/librustc/ty/sty.rs:
```diff
@@ -370,33 +364,42 @@ pub struct ClosureSubsts<'tcx> {
-        let SplitClosureSubsts { upvar_kinds, .. } = self.split(def_id, tcx);
-        upvar_kinds.iter().map(|t| {
+    pub fn upvar_tys(self) -> impl Iterator<Item = Ty<'tcx>> + 'tcx {
+        let upvars = match self.split().tupled_upvars_ty.kind {
```
Heh, I added `is_valid` to side-step exposing the tuple type when it might not actually be `Tuple(..)` (debug logging hits this due to `tcx.type_of(closure_def_id)`, shortly before it gets substituted with the right types).
`opt_upvar_tys` is a good compromise, I suppose.

eddyb commented on src/librustc_infer/infer/mod.rs:
```diff
@@ -1510,8 +1506,11 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
-        let closure_sig_ty = substs.as_closure().sig_ty(def_id, self.tcx);
+    // FIXME(eddyb) this should be unnecessary, as the shallowly resolved
+    // type is known at the time of the creation of `ClosureSubsts`,
+    // see `rustc_typeck::check::closure`.
```
Yes, see #70089.

eddyb commented on src/librustc_infer/infer/type_variable.rs:
```diff
@@ -54,6 +54,7 @@ pub enum TypeVariableOriginKind {
 
     /// One of the upvars or closure kind parameters in a `ClosureSubsts`
     /// (before it has been determined).
+    // FIXME(eddyb) distinguish upvar inference variables from the rest.
```
Potentially better suggestions? Unsure where this reaches diagnostics, if anywhere, tbh.

nikomatsakis replied:
OK, I just wondered if there was something specific you had in mind.

bors commented:
:umbrella: The latest upstream changes (presumably #70211) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit d9a15ccd976bc8a4a79c99fbf68892c2bbdfffec has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis d9a15ccd976bc8a4a79c99fbf68892c2bbdfffec -->
<!-- homu: {"type":"Approved","sha":"d9a15ccd976bc8a4a79c99fbf68892c2bbdfffec","approver":"nikomatsakis"} -->
